src/pgp/gpg.c

```diff
@@ -769,6 +750,198 @@ p_gpg_format_fp_str(char* fp)
+        goto out;
+    }
+
+    gpgme_data_t key_data;
+    error = gpgme_data_new_from_mem(&key_data, buffer, strlen(buffer), 1);
+    if (error != GPG_ERR_NO_ERROR) {
+        log_error("GPG: Error creating GPGME data from buffer");
+        goto out;
+    }
+
+    error = gpgme_op_import(ctx, key_data);
+
+    gpgme_data_release(key_data);
+
+    if (error != GPG_ERR_NO_ERROR) {
+        log_error("GPG: Error importing key data");
+        goto out;
+    }
+
+    gpgme_import_result_t import_result = gpgme_op_import_result(ctx);
+    gpgme_import_status_t status = import_result->imports;
+    gboolean is_valid = (status && status->result == GPG_ERR_NO_ERROR);
+
+    if (!is_valid) {
+        log_error("GPG: Error importing PGP key (%s).", status ? gpgme_strerror(import_result->imports->result) : "");
+        goto out;
+    }
+
+    gpgme_key_t key = NULL;
+    error = gpgme_get_key(ctx, status->fpr, &key, 0);
+    if (error != GPG_ERR_NO_ERROR) {
+        log_error("GPG: Unable to find imported PGP key (%s).", gpgme_strerror(error));
+        goto out;
+    }
+
+    result = _gpgme_key_to_ProfPGPKey(key);
+    gpgme_key_release(key);
+
+out:
+    gpgme_release(ctx);
+    return result;
+}
+
+/**
+ * Converts a GPGME key struct to a ProfPGPKey struct.
+ *
+ * @param key The GPGME key struct to convert.
+ * @return A newly allocated ProfPGPKey struct populated with the converted data,
+ *         or NULL if an error occurs.
+ *
+ * @note The caller is responsible for freeing the memory of the returned ProfPGPKey structure
+ *       by calling p_gpg_free_key() to avoid resource leaks.
+ */
+static ProfPGPKey*
+_gpgme_key_to_ProfPGPKey(gpgme_key_t key)
+{
+    if (key == NULL) {
+        return NULL;
+    }
+
+    ProfPGPKey* p_pgpkey = p_gpg_key_new();
+    gpgme_subkey_t sub = key->subkeys;
+    p_pgpkey->id = strdup(sub->keyid);
+    p_pgpkey->name = strdup(key->uids->uid);
+    p_pgpkey->fp = strdup(sub->fpr);
+
+    while (sub) {
+        if (sub->can_encrypt)
+            p_pgpkey->encrypt = TRUE;
+        if (sub->can_authenticate)
+            p_pgpkey->authenticate = TRUE;
+        if (sub->can_certify)
+            p_pgpkey->certify = TRUE;
+        if (sub->can_sign)
+            p_pgpkey->sign = TRUE;
+
+        sub = sub->next;
+    }
+    return p_pgpkey;
+}
+
+/**
+ * Convert a gpgme_data_t object to a null-terminated char* string.
+ *
+ * @param data The gpgme_data_t object to convert.
+ * @return The converted string allocated using malloc, which should be freed by the caller.
+ *         If an error occurs or the data is empty, NULL is returned and errors are written to the error log.
+ */
+static char*
+_gpgme_data_to_char(gpgme_data_t data)
+{
+    size_t buffer_size = 0;
+    char* gpgme_buffer = gpgme_data_release_and_get_mem(data, &buffer_size);
+
+    if (!gpgme_buffer) {
+        log_error("GPG: Unable to extract gpgmedata.");
+        return NULL;
+    }
+
+    char* buffer = strndup(gpgme_buffer, buffer_size);
```

Comment:
We don't return buffer size, it varies, so it doesn't make sense, as later it's only used as a null-terminated string.

Reply:
> We don't return buffer size, it varies, so it doesn't make sense

What do you mean by 'it'?

The idea behind the change is that the `strndup()` approach will take longer than the `memcpy()` one since it has to look at each byte whether it terminates the string, but that's not needed.

Reply:
applied